src/openforms/forms/tests/test_api_formsteps.py

```diff
@@ -856,6 +856,133 @@ def test_with_non_unique_fd_slugs(self):
         slugs = {form_step.slug for form_step in form_steps}
         self.assertEqual(len(slugs), 2)  # we expect two unique slugs
 
+    @patch("openforms.forms.tasks.create_form_variables_for_components.si")
+    @patch(
+        "openforms.forms.tasks.repopulate_reusable_definition_variables_to_form_variables.si"
+    )
+    @patch("openforms.forms.tasks.recouple_submission_variables_to_form_variables.si")
+    @patch("openforms.forms.tasks.chain")
+    @tag("gh-4824")
+    def test_create_form_step_schedules_task_chain_to_fix_form_variables(
+        self, mock_chain, mock_recouple, mock_repopulate, mock_create_variables
+    ):
+        """
+        Regression test for https://github.com/open-formulieren/open-forms/issues/4824
+        """
+        mock_chain_instance = mock_chain.return_value
+        mock_chain_instance.delay.return_value = None
+
+        assign_change_form_permissions(self.user)
+        form = FormFactory.create()
+        url = reverse("api:form-steps-list", kwargs={"form_uuid_or_slug": form.uuid})
+        form_definition = FormDefinitionFactory.create(
+            configuration={
+                "display": "form",
+                "components": [
+                    {
+                        "key": "lastName",
+                        "type": "textfield",
+                        "label": "Last Name",
+                    },
+                ],
+            }
+        )
+        formdefinition_detail_url = reverse(
+            "api:formdefinition-detail",
+            kwargs={"uuid": form_definition.uuid},
+        )
+        data = {
+            "formDefinition": f"http://testserver{formdefinition_detail_url}",
+            "index": 0,
+        }
+        self.client.force_login(user=self.user)
+
+        with self.captureOnCommitCallbacks(execute=True):
+            response = self.client.post(url, data=data)
+
+        mock_create_variables.assert_called_once_with(form.id)
+        mock_repopulate.assert_called_once_with(form.id)
+        mock_recouple.assert_called_once_with(form.id)
+
+        mock_chain.assert_called_once_with(
+            mock_create_variables.return_value,
+            mock_repopulate.return_value,
+            mock_recouple.return_value,
+        )
+
+        mock_chain_instance.apply_async.assert_called_once_with(countdown=60)
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(
+            FormStep.objects.filter(form_definition=form_definition).count(),
+            1,
+        )
+
+    @patch("openforms.forms.tasks.create_form_variables_for_components.si")
+    @patch(
+        "openforms.forms.tasks.repopulate_reusable_definition_variables_to_form_variables.si"
+    )
+    @patch("openforms.forms.tasks.recouple_submission_variables_to_form_variables.si")
```

Comment:
doesn't it make more sense to ensure that the `on_post_step_save` is called with the right arguments, and then separately write a unit test without any mocking/patching that calls that function with `CELERY_TASK_ALWAYS_EAGER` to check that the desired state is achieved?

I get nervous from seeing `unittest.mock` because it doesn't give me any guarantees that the code actually does what it's supposed to do, it's testing implementation details and in tests I really want to stay away from implementation details and be able to freely change/refactor them without having to touch tests.